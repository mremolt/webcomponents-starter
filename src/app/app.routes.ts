import './home/home-view.element';

export const APP_ROUTES = [
  {
    path: '/',
    component: 'sz-home-view',
  },
  {
    path: '/page2',
    component: 'sz-page2-view',
    action: async (): Promise<void> => {
      await import('./page2/page2-element');
    },
  },
];
