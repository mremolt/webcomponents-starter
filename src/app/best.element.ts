import { LitElement, TemplateResult, customElement, html, property } from 'lit-element';

import './even-better.element';

@customElement('my-best')
export class BestElement extends LitElement {
  @property() public wer = 'Welt';

  @property({ type: Object }) public person = { firstName: '', lastName: '' };

  @property({ type: Number }) public zahl = 12;

  private numbers = [1, 2, 3, 4, 5, 6, 7];

  constructor() {
    super();

    setTimeout(() => {
      // this.numbers = [9, 8, 7, 6, 5, 4, 3];
      this.numbers = [...this.numbers, 9, 8, 7, 6, 5, 4, 3];
      this.requestUpdate();
    }, 4000);
  }

  public render(): TemplateResult {
    return html`
      <h1>Headline</h1>
      <p>Hallo ${this.wer}!</p>
      <p>Die Zahl ${this.zahl + 5}!</p>

      <my-even-better .person="${this.person}" .numbers="${this.numbers}"></my-even-better>
    `;
  }
}
