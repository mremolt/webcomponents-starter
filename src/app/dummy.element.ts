export class DummyElement extends HTMLElement {
  constructor() {
    super();
    console.log(this);

    const root = this.attachShadow({ mode: 'closed' });

    const wer = 'CLEAR-IT';

    root.innerHTML = `
    <style>
      p {
        color: var(--error-color);
      }
    </style>
      <p>
        Hallo ${wer}
        Hallo Welt
        Hallo Welt
        Hallo Welt
        Hallo Welt
        Hallo Welt
        Hallo Welt
      </p>`;
  }

  public doSomething(): void {
    console.log('I am doing something', this);
  }
}

customElements.define('my-dummy-element', DummyElement);
