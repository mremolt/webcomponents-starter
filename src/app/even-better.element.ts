import { LitElement, TemplateResult, customElement, html, property } from 'lit-element';

@customElement('my-even-better')
export class EvenBetterElement extends LitElement {
  @property({ type: Object }) public person = { firstName: '', lastName: '' };

  @property({ type: Array }) public numbers: number[] = [];

  public render(): TemplateResult {
    return html`
      <p>Vorname: ${this.person.firstName}</p>
      <p>Nachname: ${this.person.lastName}</p>

      <ul>
        <li>${this.numbers[0]}</li>
        ${this.renderNumbers(this.numbers)}
      </ul>
    `;
  }

  public renderNumbers(numbers: number[]): TemplateResult[] {
    function renderLi(num: number): TemplateResult {
      return html`
        <li>${num}</li>
      `;
    }

    return numbers.map(renderLi);
  }
}
