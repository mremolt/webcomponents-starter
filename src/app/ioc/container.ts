import { IocContainer } from './ioc-container.class';

export const container = new IocContainer();
