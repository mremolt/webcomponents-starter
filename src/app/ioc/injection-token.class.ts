export class InjectionToken<T> {
  public readonly token: symbol;
  // eslint-disable-next-line
  protected _type: T;

  constructor(description: string) {
    this.token = Symbol(description);
  }
}
