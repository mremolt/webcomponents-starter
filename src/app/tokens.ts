import { InjectionToken } from './ioc/injection-token.class';
import { TranslationService } from './services/translation.service';
import { store } from './store/store';

export const REDUX_STORE = new InjectionToken<typeof store>('REDUX_STORE');
export const TRANSLATION_SERVICE = new InjectionToken<TranslationService>('TRANSLATION_SERVICE');
