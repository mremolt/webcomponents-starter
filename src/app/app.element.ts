import { CSSResult, LitElement, css, customElement, html } from 'lit-element';
import { Router } from '@vaadin/router';
import { TemplateResult } from 'lit-html';

import { APP_ROUTES } from './app.routes';
import { REDUX_STORE } from './tokens';
import { RootState } from './store/root.reducer';
import { connect } from './store/store';
import { container } from './ioc/container';
import { increment } from './store/counter/counter.actions';
import { selectValue } from './store/counter/counter.selectors';
import { withTranslations } from './mixins/translation.mixin';

import './elements/header.element';

@customElement('sz-app')
export class AppElement extends withTranslations(connect(LitElement)) {
  public static get styles(): CSSResult {
    return css`
      :host {
        display: block;
      }

      #outlet {
        margin: 0px 10px 0px 10px;
      }
    `;
  }

  private router: Router;

  private readonly store = container.inject(REDUX_STORE);

  private counter = 0;

  get outlet(): HTMLDivElement {
    return this.shadowRoot.getElementById('outlet') as HTMLDivElement;
  }

  public render(): TemplateResult {
    return html`
      <sz-header></sz-header>

      <div>Counter: ${this.counter}</div>

      <div id="outlet"></div>
    `;
  }

  public connectedCallback(): void {
    super.connectedCallback();

    setInterval(this.triggerIncrement.bind(this), 1500);
  }

  public firstUpdated(): void {
    this.router = new Router(this.outlet);
    this.setupRoutes();
  }

  public stateChanged(state: RootState): void {
    this.counter = selectValue(state);
  }

  public triggerIncrement(): void {
    this.store.dispatch(increment());
  }

  private setupRoutes(): void {
    this.router.setRoutes(APP_ROUTES);
  }
}
