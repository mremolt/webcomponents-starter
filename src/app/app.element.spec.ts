import configureStore, { MockStore } from 'redux-mock-store';

import { AppElement } from './app.element';
import { REDUX_STORE, TRANSLATION_SERVICE } from './tokens';
import { TranslationService } from './services/translation.service';
import { container } from './ioc/container';

describe('AppElement', () => {
  let subject: AppElement;
  let store: MockStore<any>;

  beforeAll(() => {
    store = configureStore()({});
    container.provide(REDUX_STORE, () => store);
    container.provide(TRANSLATION_SERVICE, () => new TranslationService());
  });

  beforeEach(() => {
    store.clearActions();
    subject = new AppElement();
  });

  it('should build', () => {
    expect(subject).toBeTruthy();
  });

  describe('render', () => {
    it('should return the template', () => {
      expect(subject.render()).toMatchSnapshot();
    });
  });

  describe('triggerIncrement', () => {
    beforeEach(() => {
      subject.triggerIncrement();
    });

    it('should call store.dispatch with increment', () => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
