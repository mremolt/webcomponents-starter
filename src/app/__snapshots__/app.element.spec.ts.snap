// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppElement render should return the template 1`] = `
TemplateResult {
  "processor": DefaultTemplateProcessor {},
  "strings": Array [
    "
      <sz-header></sz-header>

      <div>Counter: ",
    "</div>

      <div id=\\"outlet\\"></div>
    ",
  ],
  "type": "html",
  "values": Array [
    0,
  ],
}
`;

exports[`AppElement triggerIncrement should call store.dispatch with increment 1`] = `
Array [
  Object {
    "error": undefined,
    "meta": undefined,
    "payload": undefined,
    "type": "[Counter] Increment",
  },
]
`;
